### DISCUSSION TITLE
SyncCrazyflie on a live link, along with lighthouse memory

### DISCUSSION THREAD

**Discussion by willgraham345** [bitcraze/discussions#1185] [Q&A] [answered]
Hi all! As the title suggests, I'm running into issues using the `LighthouseMemHelper` within an active URI link on the crazyflie. 

I've attached a screenshot of a debugging thread, where I've attempted to use the `Event()` as demonstrated in the [example usage](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/lighthouse/read_lighthouse_mem.py). The first screenshot details my attempts (and my other thoughts), while the second screenshot shows that the `Event()` has timed out. 

My questions are:
1. Do I need to instantiate a `SyncCrazyflie()` instance to read/write the Lighthouse memory? If not, is there an alternative method or class I can use?
2. Does `SyncCrazyflie()` enable connecting to an already linked Crazyflie? 
3. If I write using the `SyncCrazyflie()` instance, will this be stored in permanent memory, or will it end when I close the link on the `SyncCrazyflie()` context?
4. Can the `cf.commander.send_hover_setpoint()` method be called wihtin a `SyncCrazyflie` instance? Could I start and stop a `SyncCrazyflie` while maintaining a hover?

Happy to provide additional code or do anything else to get this resolved. Thanks! :)

Screenshot 1:
![Screenshot from 2024-02-16 13-07-56](https://github.com/bitcraze/discussions/assets/49169367/ad3885f6-08eb-4fce-b70e-ff3bf5869e26)

Screenshot 2:
![Screenshot from 2024-02-16 13-08-12](https://github.com/bitcraze/discussions/assets/49169367/7d60e07b-ffb3-4c5b-9ef7-227353ca83af)

**Comment by ataffanel** [comment, marked as answer]
Hi,

This example does not really use the sync Crazyflie. The LighthouseMemHelper is called with ```scf.cf```. This is the normal Crazyflie object contained in the SyncCrazyflie object. This example uses the python standard Event() to help implementing things synchronously: the LighthouseMemHelper takes callbacks as parameters, and call the callback function when the work is done. Event in this example are just used to unlock the main flow of execution when the callback function are called.

I will try to answer your question in order:
 1. No you do not, LighthouseMemHelper takes a regular Crazyflie object.
 2. No, SyncCrazyflie is a helper to write simple scripts in a synchronous way. You have to create it and then you can use the crazyflie object inside it.
 3. I am not sure to understand the question. The concept of the SyncCrazyflie implementation is to close the link when you leave the synchronous block.
 4. Yes, ```scf.cf``` is the Crazyflie object, so you can call any method you want on it.

You can read a bit more about the SyncCrazyflie in the documentation: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/python_api/#synchronous-api. It is only a convenience wrapper around the Crazyflie object, so you should never be forced to use it unless you need/want to.

**Reply by willgraham345** [reply]
Thank you for the response! I realized that I was incorrectly using the `Event()` feature built within python. 

One final question for you, I've noticed that my lighthouse deck has had the yellow light blink (on and then off). I was able to find [this post](https://forum.bitcraze.io/viewtopic.php?t=3467#:~:text=How%20does%20the%20LED%20sequence%20on%20the,and%20after%20that%20red%20should%20turn%20off) regarding the issue, but it unfortunately didn't cover this particular question. As some added context, I noticed that the blinking happens only when in view of the lighthouses.

I should mention that I've done some firmware modification with one of the Lighthouse macros in the [lighthouse_position_est.c](https://github.com/bitcraze/crazyflie-firmware/blob/aefa1d6ed7c8f363d43473e02c53211be427b933/src/modules/src/lighthouse/lighthouse_position_est.c#L212) (code block below shows modifications). 
```c
// Sensor positions on the deck
#define SENSOR_POS_W (0.015f / 2.0f)
#define SENSOR_POS_L (0.030f / 2.0f)
// ORIGINAL SENSOR LAYOUT
// static vec3d sensorDeckPositions[4] = {
//     {-SENSOR_POS_L, SENSOR_POS_W, 0.0},
//     {-SENSOR_POS_L, -SENSOR_POS_W, 0.0},
//     {SENSOR_POS_L, SENSOR_POS_W, 0.0},
//     {SENSOR_POS_L, -SENSOR_POS_W, 0.0},
// };

// FLIPPED 180 DEG ABOUT  X AXIS SENSOR LAYOUT
static vec3d sensorDeckPositions[4] = {
    {-SENSOR_POS_L, -SENSOR_POS_W, 0.0},
    {-SENSOR_POS_L, SENSOR_POS_W, 0.0},
    {SENSOR_POS_L, -SENSOR_POS_W, 0.0},
    {SENSOR_POS_L, SENSOR_POS_W, 0.0},
};
```

I flashed this onto the firmware using the instructions [here](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/building-and-flashing/build/). Namely: `make menuconfig` -> `make -j 12`, -> `make cload`

**Reply by ataffanel** [reply]
Yellow blinking should mean that sweeps are received from the lighthouse basestation, this means that the system is likely not fully setup.

**Reply by willgraham345** [reply]
Great! I'm assuming by system setup, you're referring to a calibration and geometry being written to the crazyflie. Is that right?

I'll close this chat after you respond. Thank you for the help!

**Reply by willgraham345** [reply]
Took a closer look at my script and found that I indeed wasn't writing the calibration data. I'll leave this comment up if someone else is seeing a blinking yellow light on the lighthouse deck. Thanks again!!

**Comment by knmcguire** [comment]
hi! I've just reopened the discussion, since closing it will remove it from the list, and we'd like people to see the answer

**Reply by willgraham345** [reply]
Oops! I thought I merely marked it as being answered. Thanks again for the help!
